backend/gsr_booking/api_wrapper.py

```diff
@@ -532,38 +311,175 @@ def get_affiliation(self, email):
+            raise APIError(f"{str(e)}. Was only able to book {start.strftime('%H:%M')} - {curr_start.strftime('%H:%M')}")
+
+        return reservation
+    
+    def cancel_room(self, booking_id, user):
+        # I think this should prefetch reservation, reservation__gsrbooking_set, and gsr ?
+        if gsr_booking := GSRBooking.objects.filter(booking_id=booking_id).prefetch_related(Prefetch('reservation__gsrbooking_set'), Prefetch('gsr')).first():
+            if gsr_booking.user != user and gsr_booking.reservation.creator != user:
                 raise APIError("Error: Unauthorized: This reservation was booked by someone else.")
+            
+            (self.WBW.cancel_room if gsr_booking.gsr.kind == GSR.KIND_WHARTON else self.LBW.cancel_room)(booking_id, gsr_booking.user)
+            
             gsr_booking.is_cancelled = True
             gsr_booking.save()
-        response = self.request("POST", f"{API_URL}/1.1/space/cancel/{booking_id}").json()
-        if "error" in response[0]:
-            raise APIError("LibCal: " + response[0]["error"])
-        return response
 
-    def check_credits(self, user, lc_start=None):
-        # default to beginning of day
-        if lc_start is None:
-            lc_start = make_aware(datetime.datetime.now())
-            lc_start = lc_start.replace(second=0, microsecond=0, minute=0, hour=0)
-
-        lc_end = lc_start + datetime.timedelta(days=1)
-
-        # filters for all reservations for the given date
-        reservations = GSRBooking.objects.filter(
-            gsr__in=GSR.objects.filter(kind=GSR.KIND_LIBCAL),
-            start__gte=lc_start,
-            end__lte=lc_end,
-            is_cancelled=False,
-            user=user,
-        )
-        total_minutes = 0
-        for reservation in reservations:
-            # accumulates total minutes over all reservations
-            total_minutes += int((reservation.end.timestamp() - reservation.start.timestamp()) / 60)
-        # 120 minutes per day
-        return {CreditType.LIBCAL.value: 120 - total_minutes}
+            reservation = gsr_booking.reservation
+            if all(booking.is_cancelled for booking in reservation.gsrbooking_set.all()):
+                reservation.is_cancelled = True
+                reservation.save()                
+        else:
+            try: 
+                self.WBW.cancel_room(booking_id, user)
+            except APIError:
+                try: 
+                    self.LBW.cancel_room(booking_id, user)
+                except APIError:
+                    raise APIError("Error: Unknown booking id")
+                
+        
+    def get_availability(self, lid, gid, start, end, user, group=None):            
+        gsr = get_object_or_404(GSR, gid=gid)
+
+        # select a random user from the group if booking wharton gsr
+        if gsr.kind == GSR.KIND_WHARTON and group is not None:
+            wharton_members = group.memberships.filter(is_wharton=True)
```

Comment:
Hm, I feel like we don't need the additional computation for getting random. I think we can just always use the first person with (`.first()`). Maybe we would need this random so that wharton doesn't get suspicious when this one user is using this system a lot but otherwise I'm not sure